python/dgl/mock_sparse2/matmul.py

```diff
@@ -53,11 +53,94 @@ def spmm(A: Union[SparseMatrix, DiagMatrix], X: torch.Tensor) -> torch.Tensor:
     return torch.ops.dgl_sparse.spmm(A.c_sparse_matrix, X)
 
 
+def _diag_diag_mm(A1: DiagMatrix, A2: DiagMatrix) -> DiagMatrix:
+    """Internal function for multiplying a diagonal matrix by a diagonal matrix
+
+    Parameters
+    ----------
+    A1 : DiagMatrix
+        Matrix of shape (N, M), with values of shape (nnz1)
+    A2 : DiagMatrix
+        Matrix of shape (M, P), with values of shape (nnz2)
+
+    Returns
+    -------
+    DiagMatrix
+        The result of multiplication.
+    """
+    M, N = A1.shape
+    N, P = A2.shape
+    common_diag_len = min(M, N, P)
+    new_diag_len = min(M, P)
+    diag_val = torch.zeros(new_diag_len)
+    diag_val[:common_diag_len] = (
+        A1.val[:common_diag_len] * A2.val[:common_diag_len]
+    )
+    return diag(diag_val.to(A1.device), (M, P))
+
+
+def spspmm(
+    A1: Union[SparseMatrix, DiagMatrix], A2: Union[SparseMatrix, DiagMatrix]
+) -> Union[SparseMatrix, DiagMatrix]:
+    """Multiply a sparse matrix by a sparse matrix. The non-zero values of the
+    two sparse matrices must be 1D.
+
+    Parameters
+    ----------
+    A1 : SparseMatrix or DiagMatrix
+        Sparse matrix of shape (N, M) with values of shape (nnz)
+    A2 : SparseMatrix or DiagMatrix
+        Sparse matrix of shape (M, P) with values of shape (nnz)
+
+    Returns
+    -------
+    SparseMatrix or DiagMatrix
+        The result of multiplication. It is a DiagMatrix object if both matrices
+        are DiagMatrix objects. It is a SparseMatrix object otherwise.
+
+    Examples
+    --------
+
+    >>> row1 = torch.tensor([0, 1, 1])
+    >>> col1 = torch.tensor([1, 0, 1])
+    >>> val1 = torch.ones(len(row1))
+    >>> A1 = create_from_coo(row1, col1, val1)
+
+    >>> row2 = torch.tensor([0, 1, 1])
+    >>> col2 = torch.tensor([0, 2, 1])
+    >>> val2 = torch.ones(len(row2))
+    >>> A2 = create_from_coo(row2, col2, val2)
+    >>> result = dgl.sparse.spspmm(A1, A2)
+    >>> print(result)
+    SparseMatrix(indices=tensor([[0, 0, 1, 1, 1],
+                                 [1, 2, 0, 1, 2]]),
+                 values=tensor([1., 1., 1., 1., 1.]),
+                 shape=(2, 3), nnz=5)
+    """
+    assert isinstance(
+        A1, (SparseMatrix, DiagMatrix)
+    ), f"Expect A1 to be a SparseMatrix or DiagMatrix object, got {type(A1)}"
+    assert isinstance(
+        A2, (SparseMatrix, DiagMatrix)
+    ), f"Expect A2 to be a SparseMatrix or DiagMatrix object, got {type(A2)}"
+
+    if isinstance(A1, DiagMatrix) and isinstance(A2, DiagMatrix):
+        return _diag_diag_mm(A1, A2)
+    if isinstance(A1, DiagMatrix):
```

Comment:
Will the performance be improved if we have dedicated Diag @ Sparse implementation?
If so, do we plan to revisit the implementation here?

Reply:
It is possible, but it requires non-trivial implementation. Since Diag @ sparse is not frequently used, its performance improvement should not have high priority.